Add description of data node bootstrapping (#266)
timescale/docs.timescale.com-content: reference/data-node-bootstrapping.md

# Data Node Bootstrapping [](distdb-bootstrap)

In order for an access node to be able to use a data node it is

> necessary to ensure that it is properly setup to work as a data

Author: Done.

node. This is called *bootstrapping* the data node.

> When an access node adds a new data node to the distributed database

Reviewer: ```suggestion
When an access node is adding a new data node to the distributed database
```

it either bootstraps the data node or verifies that the data node is
properly bootstrapped.

> A new data node is added to the access node using the `add_data_node`

Author: Done.

command. The default database to bootstrap on the data node is then
given by the same name, that is if you are connected to the `my_data`
database on the access node, it will be `my_data` on the data node as

> well. If you do not want to use the same database on the data node as

Reviewer 2: I was confused by the formation, since it is not clear `the same name` as what, and it is unclear that `my_data` is an example.
So here is my suggestion:

> The default database to bootstrap has the same name as the access node database, i.e., the connected database from where `add_data_node` is called. For example, the name of the database is `my_test_db` on the access node and `add_data_node` is called, then a database will be bootstrap with name `my_test_db`. In the case of the database `postgres`, which is common starting point, `postgres` database will be bootstrap on the data node.

Author: Reworded.

on the access node, this can be overridden using the `database` option
to the `add_data_node` command.

All connections to the data node are made as the current user on the
access node and it is necessary that the user exists on the data node

> and that it is properly set up to handle password-less logins from the

Reviewer: Also, this link needs a footer reference

> and that it is properly set up to handle password-less logins from the

Author: Done.

access node. This means using either certificates or password files,
as described in the [Data Node Authentication Chapter][distdb-auth].

When bootstrapping the data node the access node will:

1. Ensure that there is a database on the data node.
2. Ensure that the extension is available in the database.
3. Set metadata on the data node.

## Setting up the database on the data node

First, the access node has to ensure that there is a properly
configured database on the data node by either bootstrapping a new
database, or verifying that it is correctly set up, depending on the

> value of the `bootstrap` option to `add_data_node`.

Reviewer 2: Will to be correct to be specific that it's a flag?
```suggestion
value of the `bootstrap` flag to `add_data_node`.
```

Reviewer 2: @solugebefola Thanks! I didn't think about compliance with our convention. Just thought it might be good idea to clarify that the values can be only `true` or `false`, which I've seen in docs outside Timescale.

The database name on the data node is either the name of the database
on the access node, or the value of the `database` option to
`add_data_node`.

### Bootstrapping the database

If `bootstrap` is true (which is the default), the access node will
first connect to either the `postgres` database or, if the `postgres`

> database does not exist, to the `template0` database and create the

Reviewer 2: ```suggestion
database does not exist, to the `template0` database on the data node and create new
```

database:

```
CREATE DATABASE <database>
   TEMPLATE template0 OWNER <owner>
   ENCODING <encoding> LC_COLLATE <collation>
   LC_CTYPE <character type>
```

Here, the *owner* is the user connected on the access node, the
*encoding*, *collation*, and *character type* is taken from the
current database on the access node.

Once the database is created, the access node disconnect and continue
with bootstrapping the extension. If the database already exists on

> the data node, this step is skipped.

Reviewer 2: Which step is skipped? I guess you mean the step of creating the database. Am I correct it means that there will be no error if the database exists and `bootstrap` is `true`?

Author: Correct, there will be no error.

Reviewer 2: I think it will be good which step is skipped.

Author: Elaborated.

> [Currently, there is no validation of the database as there is if

Author: Yes, these comments will be removed before committing once the actual work is done.

`bootstrap` is false. IMHO, we should add that. /Mats]

### Validation of the database

If `bootstrap` is false, the database will be validated by checking
that the encoding, character type, and collation is the same as the
current database on the access node.

## Setting up the extension

The `timescaledb` extension need to be installed in the database on
the data node, so the next step is ensuring that there is a valid
extension installed in the database by either creating the extension,
or verifying that the extension is compatible with the extension
installed on the access node.

### Bootstrapping the extension

#### Creating the schema for the extension

To install the extension, a schema is created to store the objects
that the `timescaledb` extension needs. The schema name is decided by

> picking the same schema name as on the access node and using that, so

Reviewer 2: The same schema as what? Is it the same as the current schema name during the call `add_data_node`?

Author: Elaborated.

if you have created the `timescaledb` extension in the default manner,
`public` will be used for the schema on the access node, hence
`public` will be used on the data node as well.

If the schema name for the extension on the access node is `public`,

> no new schema is created, but otherwise it is expected that the schema

Reviewer 2: What does it mean `expected`? Will it return error if the schema exists?

Author: Updated.

name does *not* exist.

[I think we should change this and not create the schema if it exists,
but otherwise proceed. /Mats]

To change the schema used on the access node and data node, you can
create the extension with the schema that you want:

```
CREATE SCHEMA magic;
CREATE EXTENSION timescaledb SCHEMA magic;
```

When validating the schema, it is necessary that the schema is owned
by the current user.

> > NOTE: This will place all the Timescale functions (for example,

Author: Done.

> `create_hypertable` in the `magic` schema), so you need to qualify
> all function references with `magic`. For example:
>
> ```
> SELECT * FROM magic.create_hypertable('conditions', 'time', 3);
> ```

#### Installing the extension

Once there is a valid schema for the extension, the extension will be
installed.

> [If the extension already exists, it is validated to ensure that it

Reviewer 2: Should it be a comment that the correct version of the extension is verified?

Author: It is not currently verified since the bootstrapping fails if the extension already exists, but I think we should add it if we allow the extension to be pre-installed.

Reviewer 2: You are right. It is checked when a connection is created, if I remember correctly.
